File: twelitePalRead.py
#!/usr/bin/python3
# coding: utf8

##########################################################################
#
# TWELITE PAL からデータを読み込み、表示
#
# usage : twelitePalRead DEVICE-NAME
#
# 2024/06/24	ambientへ送信間隔を５分（５回に１回）へ変更
#					本来はpalの設定で５分にするのだが、設定変更は外部から
#					不可の為、プログラムで対応した
#
#
#
##########################################################################

from serial import *
from sys import stdout, stdin, stderr, exit
import threading
import datetime
import ambient

# 大域変数
serialPort = None	# シリアルポート
readThread = None	# 読み出しスレッド
isExit = False		# プログラム終了フラグ

# センサーパルの種類
palIdList   = { '81' : '開閉', '82' : '環境', '83' : '動作', '84' : '通知' }
sensorList = { 	'00' : { 'name':'磁気',    'unit':'',   'datatype':'00' },
				'01' : { 'name':'温度',    'unit':'℃', 'datatype':'05' },
				'02' : { 'name':'湿度',    'unit':'％', 'datatype':'01' },
				'03' : { 'name':'照度',    'unit':'lux','datatype':'02' },
				'04' : { 'name':'加速度',  'unit':'mg', 'datatype':'15' },
				'05' : { 'name':'イベント','unit':'',   'datatype':'12' },
				'30' : { 'name':'ADC',     'unit':'mV', 'datatype':'11' }}

# for ambidata
AmbiChanel		= 79365
AmbiWriteKey    = "6234a8e3e6c7bd57"
AmbiDataTempID  = "d1"
AmbiDataHumiID  = "d2"
AmbiDataBritID  = "d3"
AmbiDataBatID   = "d4"

SendInterval:  int = 5			## 間隔
kSendInterval: int = 0			## カウンタ

ambidataHandle = ambient.Ambient(AmbiChanel,AmbiWriteKey)

#

# データパケットの表示
def printDataPacket(l):

	return True

# 受信メッセージの表示
def printPayload(l):
	global kSendInterval

	dt_now = datetime.datetime.now()

	sizeData = l[14]			# 受信データ数

	print("---DataNo 0x%04x" % (l[5] << 8 | l[6]) , dt_now.strftime('%H:%M'))
	print(" Device ID : %02x " % l[11])
	print(" LQI       : %d / %.2f [dbm]" % (l[4],(7 * l[4] - 1970) / 20.))

	index = 15
	for kData in range(sizeData):
		sensorID = format(l[index + 1],'02x')
		sensor = sensorList[sensorID]
		sensorName  = sensor['name']
		sensorValue = None
		valueUnit   = sensor['unit']

		if sensorID == '00':				# 磁気
			sensorValue = l[index + 4]
		elif sensorID == '01':				# 温度
			sensorValue = (l[index + 4] << 8 | l[index + 5]) / 100
			tempData    = sensorValue
		elif sensorID == '02':				# 湿度
			sensorValue = (l[index + 4] << 8 | l[index + 5]) / 100
			humiData    = sensorValue
		elif sensorID == '03':				# 照度
			sensorValue = l[index + 4] << 24 | l[index + 5] << 16 \
						| l[index + 6] <<  8 | l[index + 7]
			britData    = sensorValue
		elif sensorID == '04':				# 加速度
			sensorValue = "X:"+str(l[index + 4]<<8 | l[index + 5])+"mg,"  \
						+ "Y:"+str(l[index + 6]<<8 | l[index + 7])+"mg," 
#						+ "Z:"+str(l[index + 8] << 8 | l[index + 9]) + "mg"
		elif sensorID == '05':				# イベント
			sensorValue = l[index + 4]
		elif sensorID == '30':				# ADC
			sensorValue = l[index + 4] << 8 | l[index + 5]
			if l[index + 2] == 8:
				sensorName = '電源電圧'
				batData    = sensorValue
			elif l[index + 2] == 1:
				sensorName = 'ADC1'

		print(sensorName,end=":")
		print(sensorValue,end="")
		print(valueUnit,end=" /")

		index += 4 + l[index + 3]

	print(" ")
###	send ambient


	if kSendInterval == 0:
		print(" SEND !! ")
		r = ambidataHandle.send({ AmbiDataTempID : tempData, 
				AmbiDataHumiID : humiData, 
				AmbiDataBritID : britData, 
				AmbiDataBatID  : batData } )
		kSendInterval = SendInterval

	kSendInterval -= 1

	return True


# デバイスから受信データを取得し、バイト単位でリストに格納するスレッド
def readThread():
	global serialPort, isExit

	while True:
		if isExit: return		# 終了

		readData = serialPort.readline().rstrip()

		isCommand = False
		isStr	  = False

		if len(readData) > 0:
			c = readData[0]
			if isinstance(c,str):
				if c == ':': isCommand = True
				isStr = True
			else:
				if c == 58: isCommand = True

		if not isCommand: continue

		try:
			lst = {}
			if isStr:
				lst = map(ord,readData[1:].decode('hex'))
			else:
				import codecs
				s = readData[1:].decode("ascii")
				lst = codecs.decode(s, "hex_codec")

			chksum = sum(lst) & 0xff
			lst = lst[0:len(lst)-1]

			if chksum == 0:
				printPayload(lst)
			else:
				print ("checksum error")

		except:
			if len(readData) > 0:
				print ("Decord Error (%s)" % readData)

# 終了処理
def DoTerminate():
	global readThread, isExit

	# スレッド停止
	isExit = True

	print ("... quitting")
	time.sleep(0.5)		# スリープでスレッドの終了待ちをする

	exit(0)	

# メイン
if __name__ == '__main__':
	
	if len(sys.argv) != 2:
		print("Usage: %s {serial port name}" % sys.argv[0])
		exit(1)

	# open serial port
	try:
		serialPort = Serial(sys.argv[1], 115200, timeout=0.1)
		print ("open serial port: %s" % sys.argv[1])
	except:
		print ("cannot open serial port: %s" % sys.argv[1])
		exit(1)

	# 読み出しスレッドの開始
	readThread = threading.Thread(target=readThread)
	readThread.setDaemon(True)
	readThread.start()

	# stdinからの入力処理
	while True:
		try:
			sText = stdin.readline().rstrip()

			if len(sText) > 0:
				if sText[0] == 'q':
					DoTerminate()

		except KeyboardInterrupt:  # Ctrl-C
			Doterminate()
		except SystemExit:
			exit(0)
		except:
			print("... unknown exception detected")
			break
	
	exit(0)

File: twelitePalRead.cpp
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

#define SERIAL_PORT		"/dev/ttyUSB0"
#define	SERIAL_BAUDRATE	B115200

typedef struct {
			int		id;
			char	name[256];
			char	unit[16];
			int		dataType;
} S_SensorList;

#define		SizeOfSensorList	7
S_SensorList	sensorList[SizeOfSensorList] = {
		{ 0x00, "磁気",     ":",   0x00 },
		{ 0x01, "温度",     "℃",  0x05 },
		{ 0x02, "湿度",     "％",  0x01 },
		{ 0x03, "照度",     "lux", 0x02 },
		{ 0x04, "加速度",   "mg",  0x15 },
		{ 0x05, "イベント", "",    0x12 },
		{ 0x30, "ADC",      "mV",  0x11 } };

int		Readln(int handle, char *buffer, int maxSize);
void 	PrintPayload(const char *payload);

int main(int argc, char *argv[])
{
	int				serialHandler;
	struct termios	newTio,oldTio;
	char			buf[256];
	int				len;
	long			checksum;

	serialHandler = open(SERIAL_PORT,O_RDONLY);
	if (serialHandler < 0) {
		printf("Serial port open error\n");
		return -1;
	}

	ioctl(serialHandler, TCGETS, &oldTio);
	newTio = oldTio;

	cfmakeraw(&newTio);
	newTio.c_cflag = (SERIAL_BAUDRATE | CS8 | CLOCAL | CREAD);
	newTio.c_iflag = (IGNPAR);
	ioctl(serialHandler,TCSETS,&newTio);

	while (true) {
		len = Readln(serialHandler, buf, sizeof(buf));

		checksum = 0;
		for (int i = 0; i < len; i++) {
			checksum += buf[i];
		}
//		if ((checksum & 0xff) == 0) printf(" OK!\n");
//		else						printf(" NG\n");
	
		if (len > 0) {
			PrintPayload(buf);
		}
	}

	return 0;
}

/*******************************************************/
/*******************************************************/
/*******************************************************/
void PrintPayload(const char *payload) {
	time_t		t = time(NULL);
	char		dtNow[64];	
	int			sizeOfPayload;
	int			sizeData;
	int			idx;
	int			sensorIdx,sensorID;
	char		sensorName[256],sensorVal[256],valueUnit[256];

	strftime(dtNow,sizeof(dtNow),"%H:%M",localtime(&t));
	printf("---DataNo %04x %s\n",payload[5] << 8 | payload[6],
		dtNow);

	printf(" Device ID : %02x ",payload[11]);
	printf(" LQI : %d / %.2f [dbm]\n",payload[4],
						(7 * payload[4] - 1970) / 20.);

	idx = 14;
	sizeData = payload[idx++];
//printf("data size:%d\n",sizeData);

	for (int i = 0; i < sizeData; i++ ) {
		for (sensorIdx = 0; sensorIdx < SizeOfSensorList; sensorIdx++) {
			if (payload[idx + 1] == sensorList[sensorIdx].id)
				break;
		}
		if (sensorIdx >= SizeOfSensorList) {
			printf("unknow sennsorID\n");
			break;
		}
		sensorID = sensorList[sensorIdx].id;
		strcpy(sensorName,sensorList[sensorIdx].name);
		strcpy(valueUnit,sensorList[sensorIdx].unit);

		switch (sensorID) {
			case 0x00:					// 磁気
					sprintf(sensorVal,"%d",payload[idx + 4]);
					break;
			case 0x01:					// 温度
					sprintf(sensorVal,"%.2lf",(payload[idx + 4] << 8 
								|  payload[idx + 5]) / 100.);
					break;
			case 0x02:					// 湿度
					sprintf(sensorVal,"%.2lf",(payload[idx + 4] << 8 
								|  payload[idx + 5]) / 100.);
					break;
			case 0x03:					// 照度
					sprintf(sensorVal,"%d",payload[idx + 4] << 24 
								|  payload[idx + 5] << 16
								|  payload[idx + 6] <<  8
								|  payload[idx + 7]);
					break;
			case 0x04:					// 加速度
					sprintf(sensorVal,"X:%dmg,Y:%dmg,Z:%dmg",
						payload[idx + 4] << 8 | payload[idx + 5],
						payload[idx + 6] << 8 | payload[idx + 7],
						payload[idx + 8] << 8 | payload[idx + 9]);
					break;
			case 0x05:					// イベント
					sprintf(sensorVal,"%d",payload[idx+4]);
					break;
			case 0x30:					// ADC
					sprintf(sensorVal,"%d",payload[idx + 4] << 8
								| payload[idx + 5]);
					if (payload[idx + 2] == 8)
						strcpy(sensorName,"電源電圧");
					else 			//if (payload[idx + 2] == 1)
						strcpy(sensorName,"ADC1");
					break;
			default:
//					printf("error");
					break;
		}

		printf("%s:%s%s / ",sensorName,sensorVal,valueUnit);
		idx += 4 + payload[idx + 3];
	}
	printf("\n");
}


/*******************************************************/
/*******************************************************/
/*******************************************************/
int Readln(int handle, char *buffer, int maxSize) {
	int		retVal = 0;
	char	tmpBuf[maxSize];
	char	*pBuf;

	pBuf = buffer;
	while (true) {
		retVal = read(handle,tmpBuf,maxSize);

		if (retVal > 0) {
			for (int i = 0; i < retVal; i++) {
				if (tmpBuf[i] == 0x0d || tmpBuf[i] == 0x0a) {
					goto EXIT;
				}
				*pBuf++ = tmpBuf[i];
			}
		}
	}

EXIT:
	*pBuf = 0x00;

	// ASCII to BINARY
	retVal = pBuf - buffer;
	pBuf = buffer;
	for (int i = 1; i < retVal; i += 2) {
		*pBuf  = (buffer[i] - 0x30) > 10 ? buffer[i] - 0x41 + 10 
				: buffer[i] - 0x30;
		*pBuf  = *pBuf << 4;
		*pBuf += (buffer[i+1] - 0x30) > 10 ? buffer[i+1] - 0x41 + 10 
				: buffer[i+1] - 0x30;
		pBuf++;
	}
	*pBuf = 0x00;

	return pBuf - buffer;
}
